src/Rules/Documentation.hs

```diff
@@ -36,12 +119,72 @@ buildPackageDocumentation context@Context {..} = when (stage == Stage1) $ do
         let haddockWay = if dynamicPrograms then dynamic else vanilla
         build $ target (context {way = haddockWay}) Haddock srcs [file]
 
-    when (package == haddock) $ haddockHtmlLib %> \_ -> do
-        let dir = takeDirectory haddockHtmlLib
-        liftIO $ removeFiles dir ["//*"]
-        copyDirectory "utils/haddock/haddock-api/resources/html" dir
+-- | Build the haddocks for GHC's libraries
+buildLibraryDocumentation :: Rules ()
+buildLibraryDocumentation = do
+    "//" ++ htmlRoot -/- "libraries/index.html" %> \file -> do
+        haddocks <- allHaddocks
+        need haddocks
+        let libDocs = filter (\x -> takeFileName x /= "ghc.haddock") haddocks
+        let iface haddock = "--read-interface="
+                            ++ (takeFileName . takeDirectory) haddock ++ ","
+                            ++ haddock
+        runBuilder Haddock ([ "--gen-index"
+                            , "--gen-contents"
+                            , "-o", takeDirectory file
+                            , "-t", "Haskell Hierarchical Libraries"
+                            , "-p", "libraries/prologue.txt" ]
+                            ++ map iface libDocs)
+                           [] [file]
+  where
+    allHaddocks = do
+        pkgs <- stagePackages Stage1
+        sequence [ pkgHaddockFile $ vanillaContext Stage1 pkg
+                 | pkg <- pkgs, isLibrary pkg, isHsPackage pkg
+                                             , pkgName pkg /= "Win32" ]
```

Comment:
You don't need to filter out the `win32` package. The `stagePackages` action returns the list of packages that are actually built, so it will contain either `win32` or `unix`, depending on the platform.

Reply:
Ah, you are right. When I was checking if I could support `Win32`, I switched to `ghcPackages` instead of `stagePackages`. I seem to have missed that fix when I reverted that.